core/store/src/flat_state.rs

```diff
@@ -667,86 +677,74 @@ pub mod store_helper {
-        catchup_status_key.extend_from_slice(&shard_id.try_to_vec().unwrap());
-        catchup_status_key
-    }
-
-    fn get_catchup_status(store: &Store, shard_id: ShardId) -> bool {
-        let status: Option<bool> =
-            store.get_ser(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id)).expect(
-                format!("Error reading catchup status for flat state for shard {shard_id}")
-                    .as_str(),
-            );
-        match status {
-            None => false,
-            Some(status) => {
-                assert!(
-                    status,
-                    "Catchup status for flat state for shard {} must be true if stored",
-                    shard_id
-                );
-                true
+        let value = match status {
+            FlatStorageCreationStatus::FetchingState(status) => {
+                let mut value = vec![FETCHING_STATE];
+                value.extend_from_slice(&status.try_to_vec().unwrap());
+                value
             }
-        }
-    }
-
-    pub fn start_catchup(store_update: &mut StoreUpdate, shard_id: ShardId) {
-        store_update
-            .set_ser(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id), &true)
-            .expect(format!("Error setting catchup status for shard {shard_id}").as_str());
-    }
-
-    pub fn finish_catchup(store_update: &mut StoreUpdate, shard_id: ShardId) {
-        store_update.delete(crate::DBCol::FlatStateMisc, &catchup_status_key(shard_id));
+            FlatStorageCreationStatus::CatchingUp(block_hash) => {
+                let mut value = vec![CATCHING_UP];
+                value.extend_from_slice(block_hash.as_bytes());
+                value
+            }
+            status @ _ => {
+                panic!("Attempted to write incorrect flat storage creation status {status:?} for shard {shard_id}");
+            }
+        };
+        store_update.set(crate::DBCol::FlatStateMisc, &creation_status_key(shard_id), &value);
     }
 
     pub fn get_flat_storage_creation_status(
         store: &Store,
         shard_id: ShardId,
     ) -> FlatStorageCreationStatus {
         match get_flat_head(store, shard_id) {
-            None => FlatStorageCreationStatus::SavingDeltas,
             Some(_) => {
-                if let Some(fetching_state_status) = get_fetching_state_status(store, shard_id) {
-                    FlatStorageCreationStatus::FetchingState(fetching_state_status)
-                } else if get_catchup_status(store, shard_id) {
-                    FlatStorageCreationStatus::CatchingUp
-                } else {
-                    FlatStorageCreationStatus::Ready
+                return FlatStorageCreationStatus::Ready;
+            }
+            None => {}
+        }
+
+        let value = store
+            .get(crate::DBCol::FlatStateMisc, &creation_status_key(shard_id))
+            .expect("Error reading status from storage");
```

Comment:
it would be nice (maybe not a part of this PR) to make this method return `Result<..>` and avoid using `.expect` and `.unwrap`

Reply:
Yeah. I'm currently a bit confused why other jobs launched in `ClientActor` just log error instead of panicking, so I will think about it later.